### ISSUE TITLE
Technology-specific build command required

### ISSUE THREAD

**Issue by sdaschner** [appsody/appsody#121]
**Is your feature request related to a problem? Please describe.**
When building the final binary with `appsody build`, we need to add the possibility to run a an additional command, ideally in the same stack as the `run` command (and others).

In order to support the variety of technology that's out there, we need to perform extra build steps, especially if folks don't use multi-stage Docker builds but a simple Docker image that only adds the finally produced binary to an image that is otherwise free from build tools.

**Describe the solution you'd like**
Having the possibility to execute an additional build step, for example defined in `APPSODY_BUILD`, that is performed in the same stack as the others and can produce output that is copied to the extracted location as well. The subsequent `docker build` can then use the produced artifacts in the resulting image.

This enables a much cleaner setup of defining all build tools in the stack image, and only production-relevant technology in the final image.

Right now this is a blocker that prevents technologies from being used in custom stacks.

**Comment by neeraj-laad** [comment]
@sdaschner Thanks for raising this issue. Can you please provide an example of this scenario?

**Comment by sdaschner** [comment]
Sure: Have a look at the following alternative version of `java-microprofile`: https://github.com/sdaschner/appsody-stacks/tree/master/incubator/java-microprofile

The `Dockerfile` for the final build does not contain any build-specific tools, such as Maven. That is, we need to add the built binary to the resulting docker image (the WAR file in my example, under `image/project/Dockerfile`). For that to happen, we need a way to pass artifacts generated from the source code as well.

Ideally, this is a step that is executed in the stack container (from `Dockerfile-stack`), since this typically includes the required build tools anyway. Then developers can keep the project Docker image clean. I am aware that we could use multi-stage builds for this, however, from my experience, most real-world projects don't use them and IMO we should support as many technologies and ways to build an app, as possible.

Desired workflow:

`appsody build` performs the following:

- Extract the project (as it's the case today)
- Run a Docker container from the stack image (`Dockerfile-stack`), similar to what happens in `appsody run` and execute what's defined in `APPSODY_BUILD`. Have a volume with produced artifacts mounted into the container.
- Copy the contents of said mounted output folder into the exported directory
- Run the `docker build`. The (target) `Dockerfile` can use the produced artifacts which reside in the Docker build context.

**Comment by BarDweller** [comment]
How does this differ from the multi-stage build approach used by the spring boot stack ?

**Comment by sdaschner** [comment]
@BarDweller The final `Dockerfile` is not a multi-stage build in this case. Appsody builds should support the possibility for simpler builds where an already existent production image can be used with only adding the production binary. This would make the adoption of Appsody easier. The workflow is close to what the majority of projects (my experience) uses atm.

**Comment by ebullient** [comment]
Right. The Spring stack as we have it requires the final image to go through our multi-stage Dockerfile.. where someone else may have a base that is easier to certify and maintain.
